### PR TITLE
net: http: dont timeout on HTTP requests w/o body

### PR DESCRIPTION
**Pull request by mike-scott**
A TCP FIN message is passed on to user apps as a tcp_received_callback
with a NULL pkt parameter.  This means the connection is closing and
the app should do whatever cleanup it needs as there will be no further
callbacks for the current TCP connection.

Currently, if a HTTP client request doesn't receive a "body" which
the HTTP parser can use to trigger on_message_complete, then the request
will end up timing out and most apps will thing an error has occurred.

Instead, let's handle the TCP FIN message and return the waiting
semaphore, leaving the app to deal with whatever has been set in the
current HTTP context response data (IE: http_status).

This fixes using HTTP client to send POST data to servers which
only respond with HTTP_OK status and no body.

Signed-off-by: Michael Scott <michael.scott@linaro.org>

### CODE REVIEW

**Comment by mike-scott** [comment]
NOTE: This solution is somewhat debatable.   I'm open to suggestions.

**Review by jukkar — APPROVED** [review]
LGTM

**Comment by pfalcon** [comment]
In response to @mike-scott's later comment, I'm not sure if there're better ways to deal with it, but the patch definitely seems to do the right thing. There're however quick-write issues:

In commit message:

> and most apps will thing an error has occurred.

Should be "think".

In code comment:

> +		 * an http request received a server response without a body
> +		 * some methods like POST, etc this is completely normal.

This is a bit hard to parse, I'd really suggest to fix that.

File: subsys/net/lib/http/http_client.c [hunk 1]
```diff
@@ -489,6 +489,19 @@ static void recv_cb(struct net_context *net_ctx, struct net_pkt *pkt,
 	}
 
 	if (!pkt || net_pkt_appdatalen(pkt) == 0) {
+		/*
+		 * pkt == NULL is a TCP_FIN message for the TCP callback
+		 * (this means the connection is now closed)
+		 * if we get here and req.wait.count is still 0 this means
+		 * an http request received a server response without a body
+		 * some methods like POST, etc this is completely normal.
```
Comment (pfalcon) [inline subsys/net/lib/http/http_client.c]: Suggested to rewrite this sentence to add commas/prepositions/conjunctions.

**Review by jukkar — CHANGES_REQUESTED** [review]
Removing my approval atm so that this is not accidentally merged until changes suggested by @pfalcon are done.

**Comment by mike-scott** [comment]
Fixed both the comment in the source and the commit description.  Thanks @pfalcon for the review!

**Review by jukkar — APPROVED** [review]
